packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
-      );
+    if (firstChild == null) {
+      size = constraints.smallest;
+      return;
     }
-  }
-
-  @override
-  void paint(PaintingContext context, Offset offset) {
-    final Canvas canvas = context.canvas;
 
-    if (_isSingleButtonRow(size.width)) {
-      _drawButtonBackgroundsAndDividersSingleRow(canvas, offset);
-    } else {
-      _drawButtonBackgroundsAndDividersStacked(canvas, offset);
+    if (!_debugHasValidConstraints(constraints)) {
+      size = constraints.smallest;
+      return;
     }
 
-    _drawButtons(context, offset);
-  }
-
-  void _drawButtonBackgroundsAndDividersSingleRow(Canvas canvas, Offset offset) {
-    // The vertical divider sits between the left button and right button (if
-    // the dialog has 2 buttons). The vertical divider is hidden if either the
-    // left or right button is pressed.
-    final Rect verticalDivider = childCount == 2 && !_isButtonPressed
-        ? Rect.fromLTWH(
-            offset.dx + firstChild!.size.width,
-            offset.dy,
-            dividerThickness,
-            math.max(
-              firstChild!.size.height,
-              lastChild!.size.height,
-            ),
-          )
-        : Rect.zero;
-
-    final List<Rect> pressedButtonRects = _pressedButtons.map<Rect>((RenderBox pressedButton) {
-      final MultiChildLayoutParentData buttonParentData = pressedButton.parentData! as MultiChildLayoutParentData;
-
-      return Rect.fromLTWH(
-        offset.dx + buttonParentData.offset.dx,
-        offset.dy + buttonParentData.offset.dy,
-        pressedButton.size.width,
-        pressedButton.size.height,
-      );
-    }).toList();
-
-    // Create the button backgrounds path and paint it.
-    final Path backgroundFillPath = Path()
-      ..fillType = PathFillType.evenOdd
-      ..addRect(Rect.fromLTWH(0.0, 0.0, size.width, size.height))
-      ..addRect(verticalDivider);
-
-    for (int i = 0; i < pressedButtonRects.length; i += 1) {
-      backgroundFillPath.addRect(pressedButtonRects[i]);
+    final double overallWidth = constraints.maxWidth;
+    if (!_useHorizontalLayout(overallWidth)) {
+      return super.performLayout();
     }
 
-    canvas.drawPath(
-      backgroundFillPath,
-      _buttonBackgroundPaint,
-    );
+    final double slotWidth = horizontalSlotWidthFor(overallWidth: overallWidth);
+    final double height = getMinIntrinsicHeight(overallWidth);
+    size = Size(overallWidth, height);
 
-    // Create the pressed buttons background path and paint it.
-    final Path pressedBackgroundFillPath = Path();
-    for (int i = 0; i < pressedButtonRects.length; i += 1) {
-      pressedBackgroundFillPath.addRect(pressedButtonRects[i]);
-    }
+    RenderBox slot = firstChild!;
+    double x = 0;
+    while (true) {
+      slot.layout(BoxConstraints.tight(Size(slotWidth, height)), parentUsesSize: true);
+      (slot.parentData! as FlexParentData).offset = Offset(x, 0);
+      x += slot.size.width;
```

Comment:
I think in RTL + horizontal mode the first button should appear on the right? I remember RenderFlex cares about `TextDirection` when horizontal.

Reply:
Yeah, it's a problem raised in https://github.com/flutter/flutter/issues/99060. I plan to address this in a future PR.